src/magic_folder/status.py

```diff
@@ -0,0 +1,202 @@
+
+from __future__ import (
+    unicode_literals,
+)
+
+import json
+import attr
+
+from zope.interface import (
+    Interface,
+    implementer,
+)
+
+from autobahn.twisted.websocket import (
+    WebSocketServerFactory,
+    WebSocketServerProtocol,
+)
+
+from twisted.application import (
+    service,
+)
+
+
+class IStatus(Interface):
+    """
+    An internal API for services to report realtime status
+    information. These don't necessarily correspond 1:1 to outgoing
+    messages from the status API.
+    """
+
+    def upload_started():
+        """
+        One or more items are now in our upload queue.
+        """
+
+    def upload_stopped():
+        """
+        No items are in the upload queue.
+        """
+
+
+class StatusProtocol(WebSocketServerProtocol):
+    """
+    Speaks the server side of the WebSocket status protocol, usually
+    mounted at /v1/status from our web API.
+
+    This is authenticated with the same Bearer token as the rest of
+    the /v1 API.
+    """
+
+    def __init__(self, status):
+        self.status = status
+        WebSocketServerProtocol.__init__(self)
+
+    def onOpen(self):
+        """
+        WebSocket API: successful handshake
+        """
+        self.status.client_connected(self)
+
+    def onClose(self, wasClean, code, reason):
+        """
+        WebSocket API: we've lost our connection for some reason
+        """
+        self.status.client_disconnected(self)
+
+    def onMessage(self, payload, isBinary):
+        """
+        WebSocket API: a message has been received from the client. This
+        should never happen in our protocol.
+        """
+        self.sendClose(
+            code=3000,
```

Comment:
We can't send error-codes under 3000, as I understand the RFC. Those are for WebSocket protocol errors.

4000-4999 said "reserved" in Autobahn comments is why I chose something in 3000-3999 -- but I see the actual RFC ( https://datatracker.ietf.org/doc/html/rfc6455) says 4000-4999 are for application use. So, I think we have to choose 4000+

Reply:
I do think that 1003 is probably appropriate

> 1003 indicates that an endpoint is terminating the connection
      because it has received a type of data it cannot accept (e.g., an
      endpoint that understands only text data MAY send this if it
      receives a binary message).

In our case, we don't accept any kind of data, so any data is data we cannot accept.

Reply:
We can't do that. That code is for WebSocket protocol use. The RFC seems pretty clear (and anyway, Autobahn also produces an error if applications try to send those codes themselves).

Reply:
> Autobahn also produces an error if applications try to send those codes themselves

This is definitely a good reason not to use it.

> The RFC seems pretty clear [...] that code is for WebSocket protocol use. 

What isn't clear in the RFC, that suggests it would reasonable to use it, is where the boundary is between protocol and application. It would seem reasonable to me that you could provide your websocket protocol implementation a way to parse incoming messages, and have it send `1003` in response to a failure to parse. Is that "websocket protocol use"? Actually,

>       Status codes in the range 1000-2999 are reserved for definition by
>      this protocol, its future revisions, and extensions specified in a
>      permanent and readily available public specification.

suggests that the *meaning* is defined by the RFC, but we as an endpoint can use them, according to those definitions. Which somewhat makes sense, given that the RFC talks about what goes on the wire, and doesn't require any particular separation of concerns in the implementation of an endpoint.

--- 
In any case, all this is moot, given that autobahn doesn't support it.

Reply:
Our status protocol is none of the things listed for that status-code range; "extensions" means one of these: https://www.iana.org/assignments/websocket/websocket.xhtml#extension-name

Reply:
No, but they are reserved *for definition* by those things, not use by just those things. We would be using as defined in the RFC.

Reply:
Stepping back from the text, if we are not (by the RFC) allow to use 1003 to reject data we can't accept, there wouldn't be anything that could. They do give a slight variation of what we'd be using it for as an example, but it was explicitly one example, and not an exhaustive list.

Reply:
Okay, well if other implementations let applications send it for application-level data rejection I guess you should file an Autobahn bug. (As to the "nothing that could" send that code, there's all kinds of reasons at just the websocket-level that an implementation might get bad data .. masking, framing, etc etc).